### DISCUSSION TITLE
How to create a new window with OwnerWindowId?

### DISCUSSION THREAD

**Discussion by zhuxb711** [microsoft/WindowsAppSDK#3680] [Q&A] [answered]
Hi !

I want to know how to create a window that with OwnerWindowId?
Seems that this parameter should be passed during AppWindow.Create(), but what is the next and where should the AppWindow placed?

Thanks.

**Comment by castorix** [comment]
For example :

                IntPtr hWnd = WinRT.Interop.WindowNative.GetWindowHandle(this);
                Microsoft.UI.WindowId windowId = Microsoft.UI.Win32Interop.GetWindowIdFromWindow(hWnd);
                Microsoft.UI.Windowing.OverlappedPresenter presenter1 = Microsoft.UI.Windowing.OverlappedPresenter.Create();
                presenter1.IsModal = true;
                Microsoft.UI.Windowing.AppWindow appWindow = Microsoft.UI.Windowing.AppWindow.Create(presenter1, windowId);
                appWindow.Show();

**Reply by zhuxb711** [reply]
> For example :
> 
> 
> 
>                 IntPtr hWnd = WinRT.Interop.WindowNative.GetWindowHandle(this);
> 
>                 Microsoft.UI.WindowId windowId = Microsoft.UI.Win32Interop.GetWindowIdFromWindow(hWnd);
> 
>                 Microsoft.UI.Windowing.OverlappedPresenter presenter1 = Microsoft.UI.Windowing.OverlappedPresenter.Create();
> 
>                 presenter1.IsModal = true;
> 
>                 Microsoft.UI.Windowing.AppWindow appWindow = Microsoft.UI.Windowing.AppWindow.Create(presenter1, windowId);
> 
>                 appWindow.Show();

However this is not about "Window" and just "AppWindow", so AppWindow could work separately without a Window? How to set the Owner of a "Window"?

**Reply by DarranRowe** [reply]
To put it simply, this is not something you should be thinking about.
Microsoft.UI.Xaml.Window does not provide a method of being created with an owner. This would indicate that this is a scenario that is not supported by the WinUI team. While it is technically possible to change the owner of the underlying HWND of a Microsoft.UI.Xaml.Window at runtime, this is something that can result in very bad things happening, including some reports of Windows bug checking. This is something so problematic that all documentation of it has been pulled from the documentation more recently.
Your best course of action is to make a feature proposal over on the Microsoft.UI.Xaml repository to allow a Microsoft.UI.Xaml.Window runtime class to be created with an owner.

**Comment by castorix** [comment, marked as answer]
> How to set the Owner of a "Window"?

It can be done with _GWL_HWNDPARENT_

Quick test with a Window (can be checked with Spy++) :

               Button btn1 = new Button()
                {
                    Content = "Click",
                    HorizontalAlignment = HorizontalAlignment.Center
                };              
                Window wnd1 = new Window()
                {
                    Content = btn1
                };
                IntPtr hWnd1 = WinRT.Interop.WindowNative.GetWindowHandle(wnd1);
                Microsoft.UI.WindowId windowId1 = Microsoft.UI.Win32Interop.GetWindowIdFromWindow(hWnd1);
                Microsoft.UI.Windowing.AppWindow appWindow = Microsoft.UI.Windowing.AppWindow.GetFromWindowId(windowId1);
                Microsoft.UI.Windowing.OverlappedPresenter presenter = appWindow.Presenter as Microsoft.UI.Windowing.OverlappedPresenter;

                IntPtr hWnd = WinRT.Interop.WindowNative.GetWindowHandle(this);
                SetWindowLong(hWnd1, GWL_HWNDPARENT, hWnd);
                presenter.IsModal = true;

                btn1.Click += (sender, e) =>
                {
                    Console.Beep(5000, 10);                
                };

                wnd1.Closed += (sender, e) =>
                {
                     EnableWindow(hWnd, true);
                };
              
                wnd1.Activate();
                appWindow.Show();

With :

        [DllImport("User32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
        public static extern bool EnableWindow(IntPtr hWnd, bool bEnable);
        
        public const int GWL_HWNDPARENT = (-8);
        
        public static IntPtr SetWindowLong(IntPtr hWnd, int nIndex, IntPtr dwNewLong)
        {
            if (IntPtr.Size == 4)
            {
                return SetWindowLongPtr32(hWnd, nIndex, dwNewLong);
            }
            return SetWindowLongPtr64(hWnd, nIndex, dwNewLong);
        }

        [DllImport("User32.dll", CharSet = CharSet.Auto, EntryPoint = "SetWindowLong")]
        public static extern IntPtr SetWindowLongPtr32(IntPtr hWnd, int nIndex, IntPtr dwNewLong);

        [DllImport("User32.dll", CharSet = CharSet.Auto, EntryPoint = "SetWindowLongPtr")]
        public static extern IntPtr SetWindowLongPtr64(IntPtr hWnd, int nIndex, IntPtr dwNewLong);

**Comment by zhuxb711** [comment]
@castorix Thanks, your code did works.
But I found that if I created two window "A" "B", set the owner of B to A, application will not exit if I close the window A directly.
